packages/core/src/shared/telemetry/telemetryService.ts

```diff
@@ -354,6 +369,67 @@ export class DefaultTelemetryService {
             }
         }
     }
+
+    /**
+     * Decides if two metrics are equal, by comparing:
+     * - `MetricDatum.MetricName`
+     * - `MetricDatum.Metadata` fields specified by `equalityKey`
+     *
+     * Note: these `MetricDatum` fields are NOT compared:
+     * - EpochTimestamp
+     * - Unit
+     * - Value
+     * - Passive
+     *
+     * @param metric1 Metric query (name, data, and equalityKey)
+     * @param metric2 Metric
+     *
+     */
+    public isMetricEqual(metric1: MetricQuery, metric2: MetricDatum): boolean {
+        if (metric1.name !== metric2.MetricName) {
+            return false
+        }
+        const equalityKey = metric1.equalityKey ?? metric2.Metadata?.map((o) => o.Value ?? '').filter((o) => !!o) ?? []
+        for (const k of equalityKey) {
+            // XXX: wrap with String() because findPendingMetric() may cast `MetricBase` to `MetadataObj`.
+            // TODO: fix this (use proper types instead of this hack).
+            const val1 = String(metric1.metadata[k])
+            // const val1 = metric1.Metadata?.find((o) => o.Key === k)?.Value
+            const val2 = metric2.Metadata?.find((o) => o.Key === k)?.Value
+            if (val1 !== val2) {
+                return false
+            }
+        }
+        return true
+    }
+
+    /**
+     * Searches current pending (not flushed) metrics for the first item "equal to" `metric`.
+     *
+     * @note The underlying metrics queue may be updated or flushed at any time while this iterates.
+     *
+     * @param name Metric name (ignored if `metric` is `MetricDatum`; required if `metric` is `MetricBase`)
+     * @param metric Metric or metadata
+     * @param equalityKey Fields that determine "equality".
+     *
+     */
+    public async findPendingMetric(
+        name: string,
+        metric: MetricDatum | MetricBase,
```

Comment:
Specifying metrics is awkward because `Metadata` is very awkward and is only used at a low-level (but it's also how the enqueued metrics are stored). Meanwhile, in app code we think of metrics as `MetricBase`, but they are stored as `MetricDatum` (and that also has useful info like `metricName`).

I will revisit this later.